core/src/main/java/com/linecorp/armeria/server/HttpServerCodec.java

```diff
@@ -0,0 +1,203 @@
+        init(new HttpServerRequestDecoder(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders,
+                                          initialBufferSize, allowDuplicateContentLengths, allowPartialChunks),
+             new HttpServerResponseEncoder());
+    }
+
+    /**
+     * Upgrades to another protocol from HTTP. Removes the {@link HttpRequestDecoder} and
+     * {@link HttpResponseEncoder} from the pipeline.
+     */
+    @Override
+    public void upgradeFrom(ChannelHandlerContext ctx) {
+        ctx.pipeline().remove(this);
+    }
+
+    private final class HttpServerRequestDecoder extends HttpRequestDecoder {
+
+        HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize);
+        }
+
+        HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+                                        boolean validateHeaders) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders);
+        }
+
+        HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+
+                                        boolean validateHeaders, int initialBufferSize) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders, initialBufferSize);
+        }
+
+        HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+                                 boolean validateHeaders, int initialBufferSize, boolean allowDuplicateContentLengths) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders, initialBufferSize,
+                  allowDuplicateContentLengths);
+        }
+
+        HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+                                 boolean validateHeaders, int initialBufferSize, boolean allowDuplicateContentLengths,
+                                 boolean allowPartialChunks) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders, initialBufferSize,
+                  allowDuplicateContentLengths, allowPartialChunks);
+        }
+
+        @Override
+        protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) throws Exception {
+            int oldSize = out.size();
+            super.decode(ctx, buffer, out);
+            int size = out.size();
+            for (int i = oldSize; i < size; i++) {
+                Object obj = out.get(i);
+                if (obj instanceof HttpRequest) {
+                    queue.add(((HttpRequest) obj).method());
+                }
+            }
+        }
+
+        @Override
+        protected HttpMessage createMessage(String[] initialLine) throws Exception {
+            return new DefaultHttpRequest(
```

Comment:
After some investigation, I found the logic that converts Netty request to Armeria headers.
https://github.com/line/armeria/blob/92e70170b560470a68ae18895d60b3689c8d98bb/core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java#L627-L649

At the very beginning, I thought that we can implements `io.netty.handler.codec.http.HttpRequest` and create our own HttpRequest implementation to replace `DefaultHttpRequest`.
But with this design, we still need `ArmeriaHttpUtil.toArmeria` some part of the logic because this class(HttpServerCodec) depends on netty's interface.

Reply:
>  I thought that we can implements io.netty.handler.codec.http.HttpRequest and create our own HttpRequest implementation to replace DefaultHttpRequest.

I think it's the way to go. 👍  We can probably create `RequestHeadersBuilder` in it and passes it in `Http1RequestDecoder`.

> we still need ArmeriaHttpUtil.toArmeria some part of the logic because this class(HttpServerCodec) depends on netty's interface.

Yeah, we can apply the logic at once in the `Http1RequestDecoder` or we can just use the logic inside of the `HttpHeaders` that is returned by `HttpMessage.headers()`. 

Does this make sense? 🤔

Reply:
Just want to double check. ✅ 

Is it what we want to improve/replace?
https://github.com/line/armeria/blob/361c115e2a8699bbe3f8d97391df3b37f9330b71/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L173-L175

Reply:
Yes, exactly. 👍 
We can define our own `DefaultHttpRequest` and return it in `createMessage`:
```java
@Override
protected HttpMessage createMessage(String[] initialLine) throws Exception {
    return new ArmeriaDefaultHttpRequest(HttpMethod.valueOf(initialLine[0]), initialLine[1]);
}

static final class ArmeriaDefaultHttpRequest implements HttpRequest {

    private final HttpMethod method;
    private final RequestHeadersBuilder builder;
    private final HttpHeaders headers;

    ArmeriaDefaultHttpRequest(HttpMethod method, String uri) {
        this.method = method;
        builder = RequestHeaders.builder(
                com.linecorp.armeria.common.HttpMethod.valueOf(method.name()), uri);
        headers = new ArmeriaHttpHeaders(builder);
    }

    @Override
    public HttpMethod method() {
        return method;
    }

    @Override
    public HttpHeaders headers() {
        return headers;
    }
    ...
}
```
We also need `ArmeriaHttpHeaders` class and it will prevents adding not allowed headers:
```java
 final class ArmeriaHttpHeaders extends HttpHeaders {

    private final RequestHeadersBuilder builder;

    ArmeriaHttpHeaders(RequestHeadersBuilder builder) {
        this.builder = builder;
    }

    @Override
    public String get(String name) {
        return builder.get(name);
    }

    @Override
    public Integer getInt(CharSequence name) {
        return builder.getInt(name);
    }

    @Override
    public HttpHeaders add(String name, Object value) {
        // We can add a logic that prevents adding invalid headers that is defined in ArmeriaHttpUtil.toArmeria()
        return builder.add(name, value);
    }
    ...
}
```

In `Http1RequestDecoder`, we will bring the `RequestHeadersBuilder` from the Netty request and add a host header if needed:
```java
if (nettyReq instanceof ArmeriaDefaultHttpRequest) {
    final RequestHeadersBuilder builder =
            ((ArmeriaDefaultHttpRequest) nettyReq).requestHeadersBuilder();
}
```

This is what I have imagined. 😉
Please let me know if it works or not.

Reply:
> Is it correct?

That's correct. 😄 

> But my concern is that implementing own ArmeriaHttpHeaders is pretty huge task. (Well, obviously this is unavoidable)

Yes, it is. But we don't have to implement all of the methods. It's only used in the pipeline so we can implement the methods that are called in the handler. 😉

Reply:
🙆‍♂️ 
Let me show you some code as soon as possible even if it's not completed!